File: Python/1269.py
from functools import cache
class Solution:
    def numWays(self, steps: int, arrLen: int) -> int:
        
        @cache
        def dfs(xPos, steps):
            if steps == xPos and xPos == 0:
                return 1
            if xPos > steps:
                return 0
            if xPos < 0 or xPos >= arrLen:
                return 0
            
            tmp = 0
            tmp += dfs(xPos-1, steps-1)
            tmp += dfs(xPos, steps-1)
            tmp += dfs(xPos+1, steps-1)
        
            return tmp % MOD

        MOD = 10**9+7
        return dfs(0, steps)

# Time complexity: O(steps*arrLen)
# Space complexity: O(steps*arrLen)

File: C++/1269.cpp
class Solution {
public:
    const int MOD = 1e9+7;
    int numWays(int steps, int arrLen) {
        std::vector<std::vector<long>> dp(steps + 1, std::vector<long>(std::min(arrLen, steps) + 1, -1));
        return dfs(0, steps, arrLen, dp);
    }
    int dfs(int idx, int steps, int arrLen, std::vector<std::vector<long>>& dp){
        if(steps == idx && idx == 0) return 1;
        if(idx < 0 || idx >= arrLen || idx > steps) return 0;
        if (dp[steps][idx] != -1) return dp[steps][idx];
        
        long tmp = dfs(idx, steps-1, arrLen, dp); // stay
        tmp += dfs(idx-1, steps-1, arrLen, dp); // left
        tmp += dfs(idx+1, steps-1, arrLen, dp); // right

        dp[steps][idx] = tmp % MOD;
        return dp[steps][idx];

    }
};
